rclc/src/rclc/executor.c

```diff
@@ -1067,6 +1501,127 @@ _rclc_execute(rclc_executor_handle_t * handle)
       //   break;
 
+      case ACTION_CLIENT:
+        // TODO(pablogs9): Handle action client status
+        if (handle->action_client->goal_response_available) {
+          rclc_action_goal_handle_t * goal_handle;
+          while (goal_handle =
+            rclc_action_find_first_handle_with_goal_response(handle->action_client),
+            NULL != goal_handle)
+          {
+            goal_handle->available_goal_response = false;
+            handle->action_client->goal_callback(
+              goal_handle, goal_handle->goal_accepted,
+              handle->callback_context);
+            if (!goal_handle->goal_accepted ||
+              RCL_RET_OK != rclc_action_send_result_request(goal_handle))
+            {
+              rclc_action_put_goal_handle(handle->action_client, goal_handle);
+            } else {
+              goal_handle->status = GOAL_STATE_ACCEPTED;
+            }
+          }
+        }
+        if (handle->action_client->feedback_available) {
+          rclc_action_goal_handle_t * goal_handle;
+          while (goal_handle = rclc_action_find_first_handle_with_feedback(handle->action_client),
+            NULL != goal_handle)
+          {
+            goal_handle->available_feedback = false;
+            handle->action_client->feedback_callback(
+              goal_handle,
+              handle->action_client->ros_feedback,
+              handle->callback_context);
+          }
+        }
+        if (handle->action_client->cancel_response_available) {
+          rclc_action_goal_handle_t * goal_handle;
+          while (goal_handle =
+            rclc_action_find_first_handle_with_cancel_response(handle->action_client),
+            NULL != goal_handle)
+          {
+            goal_handle->available_cancel_response = false;
+            if (handle->action_client->cancel_callback != NULL) {
+              handle->action_client->cancel_callback(
+                goal_handle,
+                goal_handle->goal_cancelled,
+                handle->callback_context);
+            }
+          }
+        }
+        if (handle->action_client->result_response_available) {
+          rclc_action_goal_handle_t * goal_handle;
+          while (goal_handle =
+            rclc_action_find_first_handle_with_result_response(handle->action_client),
+            NULL != goal_handle)
+          {
+            goal_handle->available_result_response = false;
+            handle->action_client->result_callback(
+              goal_handle,
+              handle->action_client->ros_result_response,
+              handle->callback_context);
+            rclc_action_put_goal_handle(handle->action_client, goal_handle);
+          }
+        }
+        break;
+
+      case ACTION_SERVER:
+        if (handle->action_server->goal_ended) {
+          rclc_action_goal_handle_t * goal_handle;
+          while (goal_handle =
+            rclc_action_find_first_terminated_handle(handle->action_server),
```

Comment:
`rclc_action_find_first_terminated_handle` checks for 
` if (handle->status > GOAL_STATE_CANCELING)` but I don't see where `handle->status` is updated within the while loop. Can you make sure, that you don't end up in endless while loop?

Reply:
(have you sufficient test cases to test that all these while - statements always terminate?)

Reply:
Its updated inside of `rclc_action_put_goal_handle`, which removes the handle from `handle->action_server->used_goal_handles`

This way `rclc_action_find_first_terminated_handle` will always run out of goals to check on `used_goal_handles`.